build.rs

```diff
@@ -42,44 +41,103 @@ fn main() {
+            let entry = entry?;
+            let ty = entry.file_type()?;
+            if ty.is_file() {
+                if predicate(&entry.file_name()) {
+                    return Ok(true)
+                }
+            } else if ty.is_dir() {
+                dirs.push(entry.path())
+            }
+            // Ignore symlinks
+        }
+    }
+    Ok(false)
+}
+
+
+fn main() {
+    // First, we build the SUNDIALS library, with requested modules with CMake
+
+    let mut lib_loc;
+    let mut inc_dir;
+    let mut library_type = "dylib";
     if cfg!(feature = "build_libraries") {
-        let dst = cmake::Config::new("vendor")
-            .define("CMAKE_INSTALL_LIBDIR", "lib")
-            .define("BUILD_STATIC_LIBS", static_libraries)
-            .define("BUILD_SHARED_LIBS", shared_libraries)
-            .define("BUILD_TESTING", "OFF")
-            .define("EXAMPLES_INSTALL", "OFF")
-            .define("EXAMPLES_ENABLE_C", "OFF")
-            .define("BUILD_ARKODE", feature!("arkode"))
-            .define("BUILD_CVODE", feature!("cvode"))
-            .define("BUILD_CVODES", feature!("cvodes"))
-            .define("BUILD_IDA", feature!("ida"))
-            .define("BUILD_IDAS", feature!("idas"))
-            .define("BUILD_KINSOL", feature!("kinsol"))
-            .define("OPENMP_ENABLE", feature!("nvecopenmp"))
-            .define("PTHREAD_ENABLE", feature!("nvecpthreads"))
-            .build();
-        let dst_disp = dst.display();
-        lib_loc = Some(format!("{}/lib", dst_disp));
-        inc_dir = Some(format!("{}/include", dst_disp));
+        (lib_loc, inc_dir, library_type) = build_vendored_sundials();
     } else {
         lib_loc = get_env_var("SUNDIALS_LIBRARY_DIR");
         inc_dir = get_env_var("SUNDIALS_INCLUDE_DIR");
     }
 
+    if lib_loc.is_none() && inc_dir.is_none() {
+        // No path specified, try to detect if the library is present
+        // on the system.
+        #[cfg(target_family = "unix")] {
+            // Try to find Sundials header files.
+            let std_inc = ["/usr/local/include/", "/usr/include/",
+                           "/usr/include/x86_64-linux-gnu",
+                           "/usr/local/Cellar"];
+            let predicate = |s: &OsString| {
+                s == "nvector_serial.h" };
+            let found = std_inc.iter().any(|d| {
+                is_in_dir(d, predicate).unwrap() });
+            if ! found {
+                (lib_loc, inc_dir, library_type) = build_vendored_sundials();
+            }
+        }
+        #[cfg(target_family = "windows")] {
+            let vcpkg = vcpkg::Config::new()
+                .emit_includes(true)
+                .find_package("sundials");
+            if vcpkg.is_err() {
+                (lib_loc, inc_dir, library_type) = build_vendored_sundials();
+            }
+        }
+        #[cfg(target_family = "wasm")] {
+            (lib_loc, inc_dir, library_type) = build_vendored_sundials();
+        }
```

Comment:
I tried to cross compile to wasm to no avail. That being said, I'm no wasm expert and it doesn't work in current master. Could you test wasm builds?

Reply:
I'll start by adding WASM to the CI.  Hopefully [cmake](https://crates.io/crates/cmake) supports it otherwise that would be the first step...

Reply:
What is your experience with WASM?  On my machine, `clang` does not succeed to compile the C code — but clearly, the path to `stdio.h` (present in many sundials files) is probably not the one WASM wants... (like you, I'm far from being an expert in WASM).
```
   ...
   Compiling sundials-sys v0.2.3 (/home/trch/software/Rust/Scientific/sundials-sys)
error: failed to run custom build command for `sundials-sys v0.2.3 (/home/trch/software/Rust/Scientific/sundials-sys)`

Caused by:
  process didn't exit successfully: `/home/trch/software/Rust/Scientific/sundials-sys/target/debug/build/sundials-sys-95571e3a631130d5/build-script-build` (exit status: 101)
  --- stdout
  CMAKE_TOOLCHAIN_FILE_wasm32-unknown-unknown = None
  CMAKE_TOOLCHAIN_FILE_wasm32_unknown_unknown = None
  TARGET_CMAKE_TOOLCHAIN_FILE = None
  CMAKE_TOOLCHAIN_FILE = None
  CMAKE_GENERATOR_wasm32-unknown-unknown = None
  CMAKE_GENERATOR_wasm32_unknown_unknown = None
  TARGET_CMAKE_GENERATOR = None
  CMAKE_GENERATOR = None
  CMAKE_PREFIX_PATH_wasm32-unknown-unknown = None
  CMAKE_PREFIX_PATH_wasm32_unknown_unknown = None
  TARGET_CMAKE_PREFIX_PATH = None
  CMAKE_PREFIX_PATH = None
  CMAKE_wasm32-unknown-unknown = None
  CMAKE_wasm32_unknown_unknown = None
  TARGET_CMAKE = None
  CMAKE = None
  running: "cmake" "/home/trch/software/Rust/Scientific/sundials-sys/vendor" "-DCMAKE_INSTALL_LIBDIR=lib" "-DBUILD_STATIC_LIBS=OFF" "-DBUILD_SHARED_LIBS=ON" "-DBUILD_TESTING=OFF" "-DEXAMPLES_INSTALL=OFF" "-DEXAMPLES_ENABLE_C=OFF" "-DBUILD_ARKODE=ON" "-DBUILD_CVODE=ON" "-DBUILD_CVODES=OFF" "-DBUILD_IDA=ON" "-DBUILD_IDAS=OFF" "-DBUILD_KINSOL=ON" "-DOPENMP_ENABLE=OFF" "-DPTHREAD_ENABLE=OFF" "-DCMAKE_INSTALL_PREFIX=/home/trch/software/Rust/Scientific/sundials-sys/target/wasm32-unknown-unknown/debug/build/sundials-sys-0f90bc14c5fa37e0/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC --target=wasm32-unknown-unknown" "-DCMAKE_C_COMPILER=/usr/bin/clang" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC --target=wasm32-unknown-unknown" "-DCMAKE_CXX_COMPILER=/usr/bin/clang" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC --target=wasm32-unknown-unknown" "-DCMAKE_ASM_COMPILER=/usr/bin/clang" "-DCMAKE_BUILD_TYPE=Debug"
  -- The C compiler identification is Clang 14.0.6
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - failed
  -- Check for working C compiler: /usr/bin/clang
  -- Check for working C compiler: /usr/bin/clang - broken
  -- Configuring incomplete, errors occurred!
   ...
--- stderr
  /usr/include/stdio.h:27:10: fatal error: 'bits/libc-header-start.h' file not found
  /usr/include/stdio.h:27:10: fatal error: 'bits/libc-header-start.h' file not found, err: true
  CMake Error at /usr/share/cmake-3.25/Modules/CMakeTestCCompiler.cmake:70 (message):
    The C compiler

      "/usr/bin/clang"

    is not able to compile a simple test program.
   ...
```

Reply:
As I understand it, because the sundials code expects some system resources, [Emscripten](https://emscripten.org/) should be used to compile it.  A [patch was merged](https://github.com/rust-lang/cmake-rs/issues/127#issuecomment-1055547244) on May 11 in the cmake crate for that.  So, if I do
```
rustup target add wasm32-unknown-emscripten
cargo build --target wasm32-unknown-emscripten
```
I get
```
  CMake Error at cmake/macros/SundialsCMakeMacros.cmake:89 (message):
    ************************************************************************

    ERROR: No integer type of size 8 was found.

                     Tried int64_t__int64longlonglong.
                     Try setting the advanced option SUNDIALS_INDEX_TYPE.

    ************************************************************************
  Call Stack (most recent call first):
    cmake/SundialsIndexSize.cmake:45 (print_error)
    cmake/SundialsSetupCompilers.cmake:21 (include)
    CMakeLists.txt:161 (include)


  thread 'main' panicked at '
  command did not execute successfully, got: exit status: 1
```

Reply:
There may not be a way around that: https://github.com/emscripten-core/emscripten/issues/11140#issuecomment-627584859

Reply:
See also [JavaScript BigInt to WebAssembly i64 integration](https://github.com/WebAssembly/proposals/issues/7).